File: app/services/telegram_client.rb
require "cgi"
require "json"
require "net/http"
require "tempfile"
require "tmpdir"
require "uri"

class TelegramClient
  TYPING_INTERVAL_SECONDS = 4
  MAX_SUGGESTED_REPLIES = 3

  def initialize(bot_token: AppConfig.fetch.telegram_bot_token)
    @bot_token = bot_token
  end

  def download_file_to_temp(file_id)
    file = get_file(file_id)
    file_path = file.fetch("file_path")
    temp_dir = Dir.mktmpdir("telegram-codex-file-")
    output_path = File.join(temp_dir, File.basename(file_path))
    response = Net::HTTP.get_response(URI("https://api.telegram.org/file/bot#{@bot_token}/#{file_path}"))
    raise "Failed to download Telegram file: #{response.code} #{response.message}" unless response.is_a?(Net::HTTPSuccess)

    File.binwrite(output_path, response.body)
    output_path
  end

  def send_message(chat_id, text, suggested_replies: [], remove_keyboard: false)
    params = {
      chat_id: chat_id,
      text: format_telegram_message(text),
      parse_mode: "HTML"
    }
    reply_markup = build_reply_markup(suggested_replies, remove_keyboard: remove_keyboard)
    params[:reply_markup] = JSON.generate(reply_markup) if reply_markup.present?
    post_form("sendMessage", params)
  end

  def send_chat_action(chat_id, action)
    post_form("sendChatAction", chat_id: chat_id, action: action)
  end

  def answer_callback_query(callback_query_id)
    post_form("answerCallbackQuery", callback_query_id: callback_query_id)
  end

  def edit_message_reply_markup(chat_id, message_id, suggested_replies: [])
    params = {
      chat_id: chat_id,
      message_id: message_id
    }
    reply_markup = build_reply_markup(suggested_replies)
    params[:reply_markup] = JSON.generate(reply_markup) if reply_markup.present?
    post_form("editMessageReplyMarkup", params)
  end

  def clear_message_reply_markup(chat_id, message_id)
    post_form(
      "editMessageReplyMarkup",
      chat_id: chat_id,
      message_id: message_id,
      reply_markup: JSON.generate(inline_keyboard: [])
    )
  end

  def with_typing_status(chat_id)
    begin
      send_chat_action(chat_id, "typing")
    rescue StandardError
      nil
    end

    stop = false
    thread = Thread.new do
      sleep(TYPING_INTERVAL_SECONDS)

      until stop
        begin
          send_chat_action(chat_id, "typing")
        rescue StandardError
          nil
        end
        sleep(TYPING_INTERVAL_SECONDS)
      end
    end

    yield
  ensure
    stop = true
    thread&.join(0.2)
  end

  def set_webhook(url, secret_token)
    post_form(
      "setWebhook",
      url: url,
      secret_token: secret_token,
      allowed_updates: JSON.generate([ "message" ])
    )

    Rails.logger.info("Telegram webhook configured url=#{url}")
  end

  def set_my_commands(commands)
    post_form("setMyCommands", commands: JSON.generate(commands))
  end

  private

  def get_file(file_id)
    response = Net::HTTP.get_response(URI("#{api_base}/getFile?file_id=#{CGI.escape(file_id)}"))
    raise "Telegram getFile failed: #{response.code} #{response.message}" unless response.is_a?(Net::HTTPSuccess)

    payload = JSON.parse(response.body)
    raise "Telegram getFile did not include a file path." unless payload["ok"] && payload.dig("result", "file_path").present?

    payload.fetch("result")
  end

  def post_form(method_name, params)
    uri = URI("#{api_base}/#{method_name}")
    response = Net::HTTP.post_form(uri, params)
    raise "Telegram #{method_name} failed: #{response.code} #{response.message}" unless response.is_a?(Net::HTTPSuccess)

    payload = JSON.parse(response.body)
    raise "Telegram #{method_name} returned ok=false" unless payload["ok"]

    payload["result"]
  end

  def api_base
    @api_base ||= "https://api.telegram.org/bot#{@bot_token}"
  end

  def format_telegram_message(text)
    placeholders = {}
    formatted = CGI.escapeHTML(text)
    placeholder_index = 0

    formatted = formatted.gsub(/```([\s\S]*?)```/) do
      key = placeholder_key(placeholder_index)
      placeholder_index += 1
      placeholders[key] = "<pre><code>#{::Regexp.last_match(1).strip}</code></pre>"
      key
    end

    formatted = formatted.gsub(/`([^`\n]+)`/, '<code>\\1</code>')
    formatted = formatted.gsub(/\*\*([^*\n]+)\*\*/, '<b>\\1</b>')
    formatted = formatted.gsub(/(^|\n)\#{1,6}\s+([^\n]+)/, '\\1<b>\\2</b>')

    placeholders.each do |key, value|
      formatted = formatted.gsub(key, value)
    end

    formatted
  end

  def placeholder_key(index)
    "TELEGRAM_CODE_BLOCK_#{index}__"
  end

  def build_reply_markup(suggested_replies, remove_keyboard: false)
    return { remove_keyboard: true } if remove_keyboard

    cleaned_replies = Array(suggested_replies).filter_map do |reply|
      next unless reply.is_a?(String)

      normalized_reply = reply.strip
      next if normalized_reply.empty?

      normalized_reply
    end.uniq.first(MAX_SUGGESTED_REPLIES)

    return nil if cleaned_replies.empty?

    {
      keyboard: cleaned_replies.map { |reply| [ { text: reply } ] },
      resize_keyboard: true,
      one_time_keyboard: true
    }
  end
end

File: src/main/java/com/telegram/codex/telegram/TelegramClient.java
package com.telegram.codex.telegram;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.telegram.codex.codex.ReplyParser;
import com.telegram.codex.config.AppProperties;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class TelegramClient {

    private static final Logger LOGGER = LoggerFactory.getLogger(TelegramClient.class);
    private static final int MAX_SUGGESTED_REPLIES = 3;

    private final AppProperties properties;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    private final ReplyParser replyParser;

    public TelegramClient(AppProperties properties, ObjectMapper objectMapper, ReplyParser replyParser) {
        this.properties = properties;
        this.objectMapper = objectMapper;
        this.replyParser = replyParser;
        this.httpClient = HttpClient.newHttpClient();
    }

    public Path downloadFileToTemp(String fileId) {
        Map<String, Object> file = getFile(fileId);
        String filePath = String.valueOf(file.get("file_path"));
        Path tempDir;
        try {
            tempDir = Files.createTempDirectory("telegram-codex-file-");
            Path outputPath = tempDir.resolve(Path.of(filePath).getFileName().toString());
            HttpRequest request = HttpRequest.newBuilder(URI.create("https://api.telegram.org/file/bot" + properties.getTelegramBotToken() + "/" + filePath)).GET().build();
            HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());
            if (response.statusCode() / 100 != 2) {
                throw new IllegalStateException("Failed to download Telegram file: " + response.statusCode());
            }
            Files.write(outputPath, response.body());
            return outputPath;
        } catch (Exception error) {
            throw new IllegalStateException("Failed to download Telegram file", error);
        }
    }

    public void sendMessage(String chatId, String text, List<String> suggestedReplies, boolean removeKeyboard) {
        NormalizedReply normalizedReply = normalizeOutboundReply(text, suggestedReplies);
        Map<String, Object> params = new LinkedHashMap<>();
        params.put("chat_id", chatId);
        params.put("text", formatTelegramMessage(normalizedReply.text()));
        params.put("parse_mode", "HTML");
        Map<String, Object> replyMarkup = buildReplyMarkup(normalizedReply.suggestedReplies(), removeKeyboard);
        if (replyMarkup != null) {
            params.put("reply_markup", writeJson(replyMarkup));
        }
        postForm("sendMessage", params);
    }

    public void sendChatAction(String chatId, String action) {
        postForm("sendChatAction", Map.of("chat_id", chatId, "action", action));
    }

    public <T> T withTypingStatus(String chatId, Supplier<T> action) {
        try {
            sendChatAction(chatId, "typing");
        } catch (Exception ignored) {
        }
        Thread thread = new Thread(() -> {
            try {
                while (!Thread.currentThread().isInterrupted()) {
                    TimeUnit.SECONDS.sleep(4);
                    sendChatAction(chatId, "typing");
                }
            } catch (Exception ignored) {
            }
        });
        thread.setDaemon(true);
        thread.start();
        try {
            return action.get();
        } finally {
            thread.interrupt();
        }
    }

    public void setWebhook(String url, String secretToken) {
        postForm("setWebhook", Map.of("url", url, "secret_token", secretToken));
        LOGGER.info("Telegram webhook configured url={}", url);
    }

    public void setMyCommands(List<Map<String, String>> commands) {
        postForm("setMyCommands", Map.of("commands", writeJson(commands)));
        LOGGER.info("Telegram commands updated count={}", commands.size());
    }

    private Map<String, Object> getFile(String fileId) {
        try {
            HttpRequest request = HttpRequest.newBuilder(
                URI.create(apiBase() + "/getFile?file_id=" + URLEncoder.encode(fileId, StandardCharsets.UTF_8))
            ).GET().build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (response.statusCode() / 100 != 2) {
                throw new IllegalStateException("Telegram getFile failed: " + response.statusCode());
            }
            Map<String, Object> payload = objectMapper.readValue(response.body(), new TypeReference<>() {
            });
            Map<String, Object> result = castMap(payload.get("result"));
            if (!Boolean.TRUE.equals(payload.get("ok")) || result == null || result.get("file_path") == null) {
                throw new IllegalStateException("Telegram getFile did not include a file path");
            }
            return result;
        } catch (Exception error) {
            throw new IllegalStateException("Failed to call Telegram getFile", error);
        }
    }

    private Object postForm(String methodName, Map<String, Object> params) {
        try {
            StringBuilder body = new StringBuilder();
            boolean first = true;
            for (Map.Entry<String, Object> entry : params.entrySet()) {
                if (!first) {
                    body.append('&');
                }
                first = false;
                body.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8));
                body.append('=');
                body.append(URLEncoder.encode(String.valueOf(entry.getValue()), StandardCharsets.UTF_8));
            }

            HttpRequest request = HttpRequest.newBuilder(URI.create(apiBase() + "/" + methodName))
                .header("Content-Type", "application/x-www-form-urlencoded")
                .POST(HttpRequest.BodyPublishers.ofString(body.toString()))
                .build();
            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (response.statusCode() / 100 != 2) {
                throw new IllegalStateException("Telegram " + methodName + " failed: " + response.statusCode());
            }
            Map<String, Object> payload = objectMapper.readValue(response.body(), new TypeReference<>() {
            });
            if (!Boolean.TRUE.equals(payload.get("ok"))) {
                throw new IllegalStateException("Telegram " + methodName + " returned ok=false");
            }
            return payload.get("result");
        } catch (Exception error) {
            throw new IllegalStateException("Failed to call Telegram " + methodName, error);
        }
    }

    private String apiBase() {
        return "https://api.telegram.org/bot" + properties.getTelegramBotToken();
    }

    private String formatTelegramMessage(String text) {
        return text
            .replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;");
    }

    private Map<String, Object> buildReplyMarkup(List<String> suggestedReplies, boolean removeKeyboard) {
        if (removeKeyboard) {
            return Map.of("remove_keyboard", true);
        }
        ArrayList<String> cleanedReplies = new ArrayList<>();
        for (String reply : suggestedReplies) {
            if (reply == null || reply.isBlank()) {
                continue;
            }
            String normalized = reply.trim();
            if (!cleanedReplies.contains(normalized)) {
                cleanedReplies.add(normalized);
            }
            if (cleanedReplies.size() == MAX_SUGGESTED_REPLIES) {
                break;
            }
        }
        if (cleanedReplies.isEmpty()) {
            return null;
        }
        List<List<Map<String, String>>> keyboard = cleanedReplies.stream()
            .map(reply -> List.of(Map.of("text", reply)))
            .toList();
        return Map.of(
            "keyboard", keyboard,
            "resize_keyboard", true,
            "one_time_keyboard", true
        );
    }

    private NormalizedReply normalizeOutboundReply(String text, List<String> suggestedReplies) {
        ReplyParser.ParsedReply payload = replyParser.parseReply(text);
        String normalizedText = payload.text().isBlank() ? text : payload.text();
        List<String> normalizedSuggestedReplies = suggestedReplies == null || suggestedReplies.isEmpty()
            ? payload.suggestedReplies()
            : suggestedReplies;
        return new NormalizedReply(normalizedText, normalizedSuggestedReplies);
    }

    private String writeJson(Object payload) {
        try {
            return objectMapper.writeValueAsString(payload);
        } catch (Exception error) {
            throw new IllegalStateException("Failed to serialize JSON", error);
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> castMap(Object value) {
        return value instanceof Map<?, ?> ? (Map<String, Object>) value : null;
    }

    private record NormalizedReply(String text, List<String> suggestedReplies) {
    }
}
